### PR TITLE
Include QueryTarget proto in error messages when serialization fails

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
To help us debug serialization failures, this PR logs the Proto message for queries that fail serialization.

Addresses https://github.com/firebase/firebase-ios-sdk/issues/6721

### CODE REVIEW

File: Firestore/core/test/unit/local/local_serializer_test.cc [hunk 1]
```diff
@@ -352,6 +352,34 @@ TEST_F(LocalSerializerTest, EncodesTargetData) {
...
+  v1::StructuredQuery::Order order;
+  order.mutable_field()->set_field_path(invalid_field_path);
+  order.set_direction(v1::StructuredQuery::ASCENDING);
+  *query_proto->mutable_structured_query()->add_order_by() = std::move(order);
+
+  ByteString bytes = ProtobufSerialize(invalid_target);
+  StringReader reader(bytes);
+
+  auto message = Message<firestore_client_Target>::TryParse(&reader);
+  serializer.DecodeTargetData(&reader, *message);
```
Comment (var-const) [inline Firestore/core/test/unit/local/local_serializer_test.cc:379]: Question: what does `message` look like after the call? (I don't suggest checking `message` in the test, just curious)

File: Firestore/core/test/unit/model/field_path_test.cc [hunk 2]
```diff
@@ -26,6 +28,12 @@ namespace firebase {
 namespace firestore {
 namespace model {
 
+namespace {
+FieldPath Parse(const std::string& path) {
+  return FieldPath::FromServerFormat(path).ConsumeValueOrDie();
+}
+}  // anonymous namespace
```
Comment (var-const) [inline Firestore/core/test/unit/model/field_path_test.cc:35]: Ultranit: the more common Google style convention is to just use a blank "name" for an unnamed namespace; i.e. this should be just `// namespace`, without the `anonymous` part.

File: Firestore/core/test/unit/model/field_path_test.cc [hunk 3]
```diff
@@ -26,6 +28,12 @@ namespace firebase {
 namespace firestore {
 namespace model {
 
+namespace {
+FieldPath Parse(const std::string& path) {
```
Comment (var-const) [inline Firestore/core/test/unit/model/field_path_test.cc:32]: Optional nit: I'd add blank lines after the opening brace of the namespace and before the closing brace to make it a little easier to see the function separately, but it's just a personal preference -- feel free to ignore.

**Review by var-const — COMMENTED** [review]
Almost LGTM.

File: Firestore/core/src/remote/serializer.h [hunk 4]
```diff
@@ -171,7 +171,7 @@ class Serializer {
 
   static google_firestore_v1_DocumentMask EncodeFieldMask(
       const model::FieldMask& mask);
-  static model::FieldMask DecodeFieldMask(
+  static util::StatusOr<model::FieldMask> DecodeFieldMask(
```
Comment (wilhuff) [inline Firestore/core/src/remote/serializer.h:174]: This method should take a `Reader` and should just return the `FieldMask`. This will significantly reduce the burden on all the callers.

This will also get rid of all the early returns you've added in the implementation. Once the reader has failed, the result of these values doesn't matter so you'll be able to get rid of the `InvalidOrderBy` method and similar--the result is already going to be invalid.

File: Firestore/core/src/remote/serializer.h [hunk 5]
```diff
@@ -189,7 +189,8 @@ class Serializer {
   EncodeListenRequestLabels(const local::TargetData& target_data) const;
 
   static pb_bytes_array_t* EncodeFieldPath(const model::FieldPath& field_path);
-  static model::FieldPath DecodeFieldPath(const pb_bytes_array_t* field_path);
+  static util::StatusOr<model::FieldPath> DecodeFieldPath(
```
Comment (wilhuff) [inline Firestore/core/src/remote/serializer.h:192]: Similarly, this should take a `Reader` and avoid returning `StatusOr`.

File: Firestore/core/test/unit/model/field_path_test.cc (hunk 3 above)
Comment (schmidt-sebastian) [reply Firestore/core/test/unit/model/field_path_test.cc:33]: Done

File: Firestore/core/test/unit/model/field_path_test.cc (hunk 2 above)
Comment (schmidt-sebastian) [reply Firestore/core/test/unit/model/field_path_test.cc]: Done. I also fixed the instance I was inspired by.

File: Firestore/core/src/remote/serializer.h (hunk 4 above)
Comment (schmidt-sebastian) [reply Firestore/core/src/remote/serializer.h]: Much cleaner. Thanks.

File: Firestore/core/src/remote/serializer.h (hunk 5 above)
Comment (schmidt-sebastian) [reply Firestore/core/src/remote/serializer.h]: Thanks

File: Firestore/core/src/remote/serializer.cc [hunk 6]
```diff
@@ -828,21 +851,28 @@ Serializer::EncodeFieldTransform(const FieldTransform& field_transform) const {
 FieldTransform Serializer::DecodeFieldTransform(
     nanopb::Reader* reader,
     const google_firestore_v1_DocumentTransform_FieldTransform& proto) const {
+  StatusOr<FieldPath> field_path = DecodeFieldPath(proto.field_path);
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply Firestore/core/src/remote/serializer.cc]: This is no longer applicable thanks to Gil's suggestion.

[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply Firestore/core/src/remote/serializer.cc]: Done

File: Firestore/core/src/model/field_path.cc [hunk 7]
```diff
@@ -154,23 +166,37 @@ FieldPath FieldPath::FromServerFormatView(absl::string_view path) {
...
+        }
         break;
 
       default:
         segment += c;
         break;
     }
     ++i;
+
+    if (!status.ok()) {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply Firestore/core/src/model/field_path.cc]: Done

File: Firestore/core/src/model/field_path.cc [hunk 8]
```diff
@@ -154,23 +166,37 @@ FieldPath FieldPath::FromServerFormatView(absl::string_view path) {
...
+      return status;
+    }
   }
-  finish_segment();
 
-  HARD_ASSERT(!inside_backticks, "Unterminated ` in path %s", path);
+  status = finish_segment();
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply Firestore/core/src/model/field_path.cc:189]: Good catch. Fixed.

File: Firestore/core/src/model/field_path.cc [hunk 9]
```diff
@@ -154,23 +166,37 @@ FieldPath FieldPath::FromServerFormatView(absl::string_view path) {
...
+
+  if (inside_backticks) {
+    status = Status{Error::kErrorInvalidArgument,
+                    StringFormat("Unterminated ` in path %s", path)};
+  }
 
-  return FieldPath{std::move(segments)};
+  return status.ok() ? StatusOr<FieldPath>{FieldPath{std::move(segments)}}
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply Firestore/core/src/model/field_path.cc]: I simplified this by using Gil's suggestion and return early in the `inside_backticks` check.

[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply Firestore/core/src/model/field_path.cc]: Done

File: Firestore/core/src/model/field_path.cc [hunk 10]
```diff
@@ -154,23 +166,37 @@ FieldPath FieldPath::FromServerFormatView(absl::string_view path) {
         break;
 
       case '\\':
-        HARD_ASSERT(i + 1 != path.size(),
-                    "Trailing escape characters not allowed in %s", path);
-        ++i;
-        segment += path[i];
+        if (i + 1 == path.size()) {
+          status =
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply Firestore/core/src/model/field_path.cc:170]: The code looks cleaner to me with just one exit point in the loop. Kept as is for now.

File: Firestore/core/test/unit/local/local_serializer_test.cc (hunk 1 above)
Comment (schmidt-sebastian) [reply Firestore/core/test/unit/local/local_serializer_test.cc:379]: `Message.ToString()` prints:

```
<Target 0x7ffeedc66090>: {
  target_id: 42
  query {
    parent: "projects/p/databases/d/documents"
    structured_query {
      from {
        collection_id: "room"
      }
      order_by {
        field {
          field_path: "`"
        }
        direction: ASCENDING
      }
    }
  }
```

I hope this bubbles up properly to Crashlytics.

File: Firestore/core/src/remote/serializer.cc [hunk 11]
```diff
@@ -1198,31 +1168,24 @@ Filter Serializer::DecodeUnaryFilter(
...
               unary.which_operand_type);
 
-  StatusOr<FieldPath> field =
-      FieldPath::FromServerFormat(DecodeString(unary.field.field_path));
-
-  if (!field.ok()) {
-    reader->set_status(field.status());
-    return InvalidFilter();
-  }
+  auto field = DecodeFieldPath(reader, unary.field.field_path);
```
Comment (var-const) [inline Firestore/core/src/remote/serializer.cc:1171]: Similar nit: similar statements above use an explicit types, please make consistent (I'm fine with either way).

File: Firestore/core/src/remote/serializer.cc [hunk 12]
```diff
@@ -1244,7 +1207,7 @@ FilterList Serializer::DecodeCompositeFilter(
   FilterList result;
   result = result.reserve(composite.filters_count);
 
-  for (pb_size_t i = 0; reader->ok() && i != composite.filters_count; ++i) {
+  for (pb_size_t i = 0; i != composite.filters_count; ++i) {
```
Comment (var-const) [inline Firestore/core/src/remote/serializer.cc]: Question: what is the reason for this change?

File: Firestore/core/src/model/field_path.cc [hunk 13]
```diff
@@ -109,23 +114,30 @@ FieldPath FieldPath::FromDotSeparatedStringView(absl::string_view path) {
...
-                path);
+    if (segment.empty()) {
+      return Status{
+          Error::kErrorInvalidArgument,
+          StringFormat(
+              "Invalid field path (%s). Paths must not be empty, begin with "
+              "'.', end with '.', or contain '..'",
+              path)};
+    }
     // Move operation will clear segment, but capacity will remain the same
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/model/field_path.cc:137]: This is unlikely to lead to an empty `FieldPath` -- if moving doesn't clear the moved-from `segment`, the resulting `FieldPath` may contain extra characters. It probably doesn't happen in practice, but still seems prudent to fix. I can do that in a follow-up PR.

File: Firestore/core/src/remote/serializer.cc (hunk 11 above)
Comment (schmidt-sebastian) [reply Firestore/core/src/remote/serializer.cc:1171]: I just want to show off my versatile coding skills here. Fixed.

File: Firestore/core/src/remote/serializer.cc (hunk 12 above)
Comment (schmidt-sebastian) [reply Firestore/core/src/remote/serializer.cc]: It simplifies the loop. The check should no longer be needed since an `InvalidFieldPath` is semantically valid in the consuming types (FieldMask, QueryTarget). I can change this back if exiting early is a better strategy here.

Comment (var-const) [reply Firestore/core/src/remote/serializer.cc]: I presume exiting early is better from the performance perspective, though now that I think about it, it only matters in the error case which should be rare. I'll leave it up to you.

**Review by var-const — APPROVED** [review]
